### PR TITLE
Implemented WPA clustering algorithm with resource balancing

### LINKED ISSUE #11: Implement WPA clustering algorithm with resource balancing

### PR DESCRIPTION
**Pull request by rpjayasekara**
This PR resolves #11 

- Implemented the WPA clustering algorithm.

- Added resource balancing to the WPA clustering algorithm.

Script path -
"swarmform/core/clustering_algo/wpa_clustering.py"

### CODE REVIEW

File: swarmform/core/cluster.py [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+    for child in children:
+        if child.get_fw_id() == child_id:
+            return True
+    return False
+
+
+def assign_parent_to_clusters(task):
+
+    """
+            Assign parent tasks of a given node to clusters
```
Comment (KalanaDananjaya) [inline swarmform/core/cluster.py]: wrong indentation level in comment

File: swarmform/core/cluster.py [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+    for parent in parents:
+        if parent.get_fw_id() == parent_id:
+            return True
+    return False
+
+
+def is_child_already_assigned(task, child_id):
+
+    """
+                    Checks weather the given node is already assigned as child or not
```
Comment (KalanaDananjaya) [inline swarmform/core/cluster.py]: wrong indentation level in comment

File: swarmform/core/cluster.py [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+    for task in cluster:
+        exec_sum += task.get_exec_time()
+    return exec_sum
+
+
+def is_parent_already_assigned(task, parent_id):
+
+    """
+                Checks weather the given node is already assigned as parent or not
+
```
Comment (KalanaDananjaya) [inline swarmform/core/cluster.py]: wrong indentation level in comment

Comment (KalanaDananjaya) [reply swarmform/core/cluster.py]: @rpjayasekara this applies to many of the comments

File: swarmform/core/cluster.py [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+                    # Delete the node from the WK
+                    wf.delete_node(parent.get_fw_id())
+                    # Assign a minus key to the refer the parallel running jobs
+                    key = -1*randrange(1, 100)
+                    # Set parallel nodes to the cluster
+                    cluster.set_parallel_ids([cluster.get_fw_ids_to_cluster_sequentially()[1],parent.get_fw_id()], key)
+                    # Set sequentially running nodes to the cluster
+                    cluster.set_sequential_ids([cluster.get_fw_ids_to_cluster_sequentially()[0], key])
+
+
```
Comment (KalanaDananjaya) [inline swarmform/core/cluster.py]: too many blank lines between the functions

File: swarmform/core/cluster.py [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+                # Add new clustered node to the WF
+                workflow.add_node(temp_wfid, cluster)
+            for cl in cls:
+                cls_at_level.append(cl)
+        # Resource balance
+        resource_balance(cls_at_level, tasks_at_level_sorted, workflow)
+    workflow.update_links()
+    return workflow
+
+
```
Comment (KalanaDananjaya) [inline swarmform/core/cluster.py]: too many blank lines

**Review by KalanaDananjaya — CHANGES_REQUESTED** [review]

File: swarmform/core/cluster.py (hunk 1 above)
Comment (rpjayasekara) [reply swarmform/core/cluster.py]: noted

File: swarmform/core/cluster.py [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+        # Iterate the tasks of the level
+        for task in tasks_at_level_sorted:
+            # Assign parents of the task to the clusters
+            cls = assign_parent_to_clusters(task)
+            cls_info = {}
+            for cluster in cls:
+                cls_info = cluster.get_cluster_info()
+                temp = 0
+                for wfid in cls_info:
+                    if temp == 0:
```
Comment (AyeshW) [inline swarmform/core/cluster.py]: Do we need to have this if block? _temp_ is always 0 and checking whether it is 0 or not

File: swarmform/core/cluster.py [hunk 7]
```diff
@@ -0,0 +1,435 @@
...
+        tasks_at_level = get_tasks_at_level(workflow, level)
+        # Sort the tasks by longest parent of the tasks
+        tasks_at_level_sorted = sort_tasks_by_longest_parent(tasks_at_level)
+        # Iterate the tasks of the level
+        for task in tasks_at_level_sorted:
+            # Assign parents of the task to the clusters
+            cls = assign_parent_to_clusters(task)
+            cls_info = {}
+            for cluster in cls:
+                cls_info = cluster.get_cluster_info()
```
Comment (AyeshW) [inline swarmform/core/cluster.py]: _cls_info_ variable at line 414 is not used as it is reassigned at line 416.

File: swarmform/core/cluster.py (hunk 5 above)
Comment (AyeshW) [reply swarmform/core/cluster.py]: need to remove ending blank lines. One line is enough

**Review by AyeshW — CHANGES_REQUESTED** [review]

File: swarmform/core/cluster.py (hunk 7 above)
Comment (rpjayasekara) [reply swarmform/core/cluster.py]: removed

File: swarmform/core/cluster.py (hunk 6 above)
Comment (rpjayasekara) [reply swarmform/core/cluster.py]: resolved that

**Review by KalanaDananjaya — APPROVED** [review]

**Review by AyeshW — APPROVED** [review]

**Pull request merged**